src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -127,6 +129,20 @@ public async Task<bool> TryCreatePendingSessionAsync(IDictionary<string, string>
 
         private Task WriteOutputMessageAsync(string outputMessage) => _hotReloadDiagnosticOutputService.Value.WriteLineAsync(outputMessage);
 
+        private async Task OnUnconfiguredProjectUnloadingAsync(object? sender, EventArgs args)
+        {
+            _project.ProjectUnloading -= OnUnconfiguredProjectUnloadingAsync;
+
+            using AsyncSemaphore.Releaser semaphoreRelease = await _semaphore.EnterAsync();
+
+            foreach (HotReloadState sessionState in _activeSessions.Values)
+            {
+                _ = sessionState.Session?.StopSessionAsync(default);
```

Comment:
Would you want to have any exceptions here logged? If so, you could use the `RunAndForget` extension on `IProjectThreadingService`.

Reply:
You don't need to hook onto Unload, just perform this in your dispose method which will be called on shutdown.

Reply:
Components who's lifetimes are tied to the project composition, shut themselves down in Dispose, implement OnceInitializedOnceDisposeAsync.

Reply:
Even better. That's the piece I was missing.